### DISCUSSION TITLE
Debounced computeds

### DISCUSSION THREAD

**Discussion by psalaets** [mobxjs/mobx#3013] [Q&A] [answered]
# Use case

I have a computed that derives its value from a rapidly changing observable (updated when user types) and the computed also has an expensive reaction (react graph). The graph doesn't need to update in real-time so I want to introduce a debounce somewhere.

I would debounce the action that updates the observable, but the observable has other reactions that *do* need to be real-time.

The description above is simplified a bit. In practice there is a chain of 2-3 computeds between the observable and the terminal reactions.

# What I'm trying

```tsx
import debounce from 'lodash/debounce';

// definition
function create<T>(getter: () => T, wait: number): IComputedValue<T> {
  const boxed = observable.box<T | undefined>(undefined, {
    deep: false
  });

  reaction(getter, debounce(value => {
    runInAction(() => boxed.set(value));
  }, wait));

  return computed(() => boxed.get()!);
}

// setting it up in a store
this.computedValue = create(() => some.burstyObservable, 1000);

// usage in an observer() component
<Graph data={store.computedValue.get()} />
```

# Questions

I'm a little wary because this violates [reaction principle #2](https://mobx.js.org/reactions.html#use-reactions-sparingly), although the docs do mention that it's not a law.

Also, there's no way to dispose the reaction. I suppose I could return `{ IComputedValue<T>, IReactionDisposer }` from `create()` to handle that.

Am I setting myself up for pain later on? Is there a cleaner way to achieve this effect?

**Comment by urugator** [comment, marked as answer]
That `computed` in `return computed(() => boxed.get()!);` is redundant.

You can either use `reaction` like you do (without computed), or you can use `computed`:
```javascript
const updateBoxed = debounce(action(value => boxed.set(value)), wait)
return computed(() => {
  updateBoxed(getter());
  return boxed.get();
})
``` 
The difference is that `reaction` is "eager" and requires disposal, while `computed` is "lazy" and suspends itself when there are no observers.
Mutating observables is (arguably) not an issue, if it does not happen synchronously (in effect or computation).

**Comment by urugator** [comment]
You may also want to check https://github.com/mobxjs/mobx-utils/issues/283
